File: src/manifest.ts
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import type { PluginManifest } from "./types.js";

export function loadManifest(path = "plugin.json"): PluginManifest {
  const manifestPath = resolve(process.cwd(), path);
  const manifest = JSON.parse(readFileSync(manifestPath, "utf8")) as PluginManifest;
  validateManifest(manifest);
  return manifest;
}

export function validateManifest(manifest: PluginManifest): void {
  if (!manifest.id) throw new Error("plugin id is required");
  if (!manifest.name) throw new Error("plugin name is required");
  if (!manifest.version) throw new Error("plugin version is required");
  if (!manifest.runtime) throw new Error("plugin runtime is required");
  if (!manifest.entry) throw new Error("plugin entry is required");
  if (!manifest.kind) throw new Error("plugin kind is required");

  for (const capability of manifest.capabilities ?? []) {
    if (!capability.name) throw new Error("capability name is required");
  }
}

File: templates/go/internal/plugin/manifest.go
package plugin

import (
	"encoding/json"
	"fmt"
	"os"
)

type Manifest struct {
	ID           string       `json:"id"`
	Name         string       `json:"name"`
	Version      string       `json:"version"`
	Runtime      string       `json:"runtime"`
	Entry        string       `json:"entry"`
	Kind         string       `json:"kind"`
	Permissions  []string     `json:"permissions,omitempty"`
	Capabilities []Capability `json:"capabilities,omitempty"`
}

type Capability struct {
	Name string `json:"name"`
}

func LoadManifest(path string) (Manifest, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return Manifest{}, err
	}

	var manifest Manifest
	if err := json.Unmarshal(data, &manifest); err != nil {
		return Manifest{}, err
	}

	if manifest.ID == "" {
		return Manifest{}, fmt.Errorf("manifest id is required")
	}
	if manifest.Name == "" {
		return Manifest{}, fmt.Errorf("manifest name is required")
	}
	if manifest.Version == "" {
		return Manifest{}, fmt.Errorf("manifest version is required")
	}
	if manifest.Runtime == "" {
		return Manifest{}, fmt.Errorf("manifest runtime is required")
	}
	if manifest.Kind == "" {
		return Manifest{}, fmt.Errorf("manifest kind is required")
	}

	return manifest, nil
}
